### PR TITLE
[cxx-interop] Support C++ default arguments

### PR DESCRIPTION
**Pull request by egorzhdan**
This allows calling a C++ function with default arguments from Swift without having to explicitly specify the values of all arguments.

For every C++ default argument we mark the corresponding Swift argument as defaulted (`= cxxDefaultArg`). If the default value of the argument is used, the compiler creates a new C++ function `__defaultArg_{argument index}_{mangled name of the function}` that returns the default expression. This function is then imported into Swift and invoked from a default expression of the imported function on the Swift side.

This patch doesn't cover default arguments of C++ constructors (https://github.com/apple/swift/issues/70124).

rdar://103975014

### CODE REVIEW

**Comment by hyp** [comment]
Could you also test things that get synthesized, e.g. base members in derived classes.

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@hyp good idea, I'll add tests for that

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please test source compatibility

**Comment by egorzhdan** [comment]
@swift-ci please test Windows

**Comment by rjmccall** [comment]
Haven't looked at the patch yet, but do we need to have constraints here to ensure that lifetimes aren't messed up?  Like, what happens if the parameter type is a reference?

File: test/Interop/Cxx/function/Inputs/default-arguments.h [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+};
+inline bool isArgZero(ArgTy a = {0}) { return a.value == 0; }
+inline bool isArgNonZero(ArgTy a = ArgTy::createNonZero()) { return a.value != 0; }
+inline bool isArgZeroConstRef(const ArgTy &a = {0}) { return a.value == 0; }
+
+struct HasMethodWithDefaultArg {
+  bool isZero(int v = 0) const { return v == 0; }
+};
+
+struct DerivedFromHasMethodWithDefaultArg : public HasMethodWithDefaultArg {};
```
Comment (hyp) [inline test/Interop/Cxx/function/Inputs/default-arguments.h:28]: Can you add a test for another level of derived (DerivedDerived) as there's another synthesis interaction happening past the first derived type.

File: test/Interop/Cxx/function/default-arguments-module-interface.swift [hunk 2]
```diff
@@ -0,0 +1,31 @@
+// RUN: %target-swift-ide-test -print-module -module-to-print=DefaultArguments -I %S/Inputs -source-filename=x -cxx-interoperability-mode=upcoming-swift | %FileCheck %s
+
+// CHECK: func isZero(_ value: Int32 = cxxDefaultArg) -> Bool
```
Comment (hyp) [inline test/Interop/Cxx/function/default-arguments-module-interface.swift:3]: is it possible to map some default arguments to literal constants instead, to present a better module interface?

File: test/Interop/Cxx/function/Inputs/default-arguments.h [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  int value;
+
+  HasCtorWithDefaultArg(int a, int b = 456, int c = 123) : value(a + b + c) {}
+};
+
+inline int ambiguous(int a, int b = 1) { return a; }
+inline int ambiguous(int a) { return a; }
+
+inline int nonTrailing(int a, int b = 2) { return a + b; }
+inline int nonTrailing(int a = 1, int b);
```
Comment (hyp) [inline test/Interop/Cxx/function/Inputs/default-arguments.h:55]: Can you add some additional tests for a class template, including a dependent value in a default parameter?

File: test/Interop/Cxx/function/default-arguments-module-interface.swift (hunk 2 above)
Comment (egorzhdan) [reply test/Interop/Cxx/function/default-arguments-module-interface.swift:3]: Good point, I was initially thinking of reusing the parsing logic in `ImportMacro.cpp` to import simple default values such as `nullptr` without synthesizing new functions. I didn't have time to implement that, but I think this would be a great future improvement.

File: test/Interop/Cxx/function/Inputs/default-arguments.h (hunk 1 above)
Comment (egorzhdan) [reply test/Interop/Cxx/function/Inputs/default-arguments.h:33]: Done

File: test/Interop/Cxx/function/Inputs/default-arguments.h (hunk 3 above)
Comment (egorzhdan) [reply test/Interop/Cxx/function/Inputs/default-arguments.h:75]: Done

**Comment by egorzhdan** [comment]
Thanks @rjmccall!

> Like, what happens if the parameter type is a reference?

We shouldn't add a default value for such parameter in Swift, since the parameter would get imported as `inout` and Swift doesn't allow default values for `inout`. This is now fixed.

I also added a test to make sure we don't add default values for foreign reference type params, we can make them work later.

Another interesting case is if a parameter has a pointer type. I think it's reasonable to assume that if the default expression allocates something, it will get deallocated by the callee. In that case we don't need to worry about lifetimes. In general, it seems importing default values of pointer parameters doesn't add extra unsafety, since the default value would refer to either a new pointer created inline or to something static/immortal.

File: lib/ClangImporter/ImportType.cpp [hunk 4]
```diff
@@ -2478,6 +2479,19 @@ static ParamDecl *getParameterInfo(ClangImporter::Implementation *impl,
...
+  // Swift doesn't support default values of inout parameters.
+  if (param->hasDefaultArg() && !isInOut &&
+      impl->isCxxInteropCompatVersionAtLeast(
+          version::getUpcomingCxxInteropCompatVersion())) {
+    SwiftDeclSynthesizer synthesizer(*impl);
+    if (CallExpr *defaultArgExpr = synthesizer.makeDefaultArgument(
+            param, swiftParamTy, paramInfo->getParameterNameLoc())) {
+      paramInfo->setDefaultArgumentKind(DefaultArgumentKind::Normal);
+      paramInfo->setDefaultExpr(defaultArgExpr, /*isTypeChecked*/ true);
+      paramInfo->setDefaultValueStringRepresentation("cxxDefaultArg");
```
Comment (hyp) [inline lib/ClangImporter/ImportType.cpp:2491]: Another idea - since using a specific string value to represent the default value in the module interface, can you instead render the original Clang expression and present it to the user as a sort of special expression , e.g. `foo(x: Foo = #cxxDefaultArg(Foo(1))`

Comment (egorzhdan) [reply lib/ClangImporter/ImportType.cpp:2491]: That is smart, I like this idea!

**Comment by hyp** [comment]
Also, please check the source compatibility test suite as well :)

**Comment by rjmccall** [comment]
> We shouldn't add a default value for such parameter in Swift, since the parameter would get imported as `inout` and Swift doesn't allow default values for `inout`. This is now fixed.

What if it's a `const &` parameter?  We don't import those as `inout`, right?

> Another interesting case is if a parameter has a pointer type. I think it's reasonable to assume that if the default expression allocates something, it will get deallocated by the callee. In that case we don't need to worry about lifetimes. In general, it seems importing default values of pointer parameters doesn't add extra unsafety, since the default value would refer to either a new pointer created inline or to something static/immortal.

I agree that pointers are probably relatively unlikely to be a problem, but view types might be dangerous, e.g. a `std::string_view` that's constructed from a temporary `std::string`.  But we should be able to detect this condition somewhat precisely, right?  The Clang AST has the actual default argument expression, and we can just look at that expression to see if it creates any temporaries and then say that such arguments can't be imported as default arguments.  (We can ignore the top-level temporary for the argument itself if the argument is a pr-value.)  This could be over-restrictive, e.g. if the default argument creates temporaries but the program doesn't depend on them staying alive across the call, but it's conservatively correct.  And it would be possible to lift that restriction in the future if we can figure out some way to extend temporaries across the call.

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please test

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please test source compatibility

**Comment by egorzhdan** [comment]
@swift-ci please test source compatibility debug

File: lib/ClangImporter/ImportType.cpp [hunk 5]
```diff
@@ -2484,6 +2486,73 @@ static ParamDecl *getParameterInfo(ClangImporter::Implementation *impl,
...
+        // creates temporaries. Since we import const T& as T in Swift, the
+        // value of the default expression will get copied in Swift unlike C++,
+        // which might be unexpected and unsafe.
+        if (param->getType()->isReferenceType() &&
+            param->getType()->getPointeeType().isConstQualified()) {
+          if (isa<clang::MaterializeTemporaryExpr>(defaultArgExpr->getExpr()))
+            isSafeToImport = false;
+        }
+      }
+    }
```
Comment (rjmccall) [inline lib/ClangImporter/ImportType.cpp:2543]: Can all this reasonably just be a function?

File: lib/ClangImporter/ImporterImpl.h [hunk 6]
```diff
@@ -516,7 +516,7 @@ class LLVM_LIBRARY_VISIBILITY ClangImporter::Implementation
   std::unique_ptr<clang::Parser> Parser;
 
   /// Clang parser, which is used to load textual headers.
-  std::unique_ptr<clang::MangleContext> Mangler;
+  std::shared_ptr<clang::MangleContext> Mangler;
```
Comment (rjmccall) [inline lib/ClangImporter/ImporterImpl.h:519]: Why is this necessary?

File: lib/IRGen/GenClangDecl.cpp [hunk 7]
```diff
@@ -54,6 +59,29 @@ class ClangDeclFinder
...
+          auto returnStmt =
+              cast<ReturnStmt>(body->getSingleActiveElement().get<Stmt *>());
+          auto callExpr = cast<CallExpr>(returnStmt->getResult());
+          auto calledFuncDecl = cast<FuncDecl>(callExpr->getCalledValue());
+          auto calledClangFuncDecl =
+              cast<clang::FunctionDecl>(calledFuncDecl->getClangDecl());
+          callback(calledClangFuncDecl);
+        }
+      }
+    }
```
Comment (rjmccall) [inline lib/IRGen/GenClangDecl.cpp:80]: So if we ever reference a function, we have to eagerly emit default argument generators for all of its parameters?  Is there some more reasonable way to make SIL carry the information for how to generate the body here so that we can just trigger this when we have a reference to the default argument generator?

File: lib/SIL/IR/SILDeclRef.cpp [hunk 8]
```diff
@@ -981,6 +981,10 @@ bool SILDeclRef::isForeignToNativeThunk() const {
   // have a foreign-to-native thunk.
   if (!hasDecl())
     return false;
+  // A default argument generator for a C++ function is a Swift function, so no
+  // thunk needed.
+  if (isDefaultArgGenerator())
+    return false;
```
Comment (rjmccall) [inline lib/SIL/IR/SILDeclRef.cpp:987]: So why is it marked as foreign?

File: lib/SILGen/SILGen.cpp [hunk 9]
```diff
@@ -719,6 +719,26 @@ static bool isEmittedOnDemand(SILModule &M, SILDeclRef constant) {
   }
   case SILDeclRef::Kind::EnumElement:
     return true;
+  case SILDeclRef::Kind::DefaultArgGenerator: {
+    // Default arguments of C++ functions are only emitted is used.
```
Comment (rjmccall) [inline lib/SILGen/SILGen.cpp:723]:
```suggestion
    // Default arguments of C++ functions are only emitted if used.
```

File: lib/SILGen/SILGen.cpp [hunk 10]
```diff
@@ -719,6 +719,26 @@ static bool isEmittedOnDemand(SILModule &M, SILDeclRef constant) {
...
+    auto funcDecl = constant.getFuncDecl();
+    if (funcDecl) {
+      // If the function is inherited from a base C++ class, we have synthesized
+      // a new Swift function that dispatches to the base method and has no
+      // clang node: the underlying clang decl is synthesized lazily. This means
+      // we can't rely on `funcDecl->hasClangNode()` here. Instead, check if any
+      // of the parameters has a clang node.
```
Comment (rjmccall) [inline lib/SILGen/SILGen.cpp:730]: That feels like something we should record about the synthesized Swift function.

**Comment by Oudinunit** [comment]
Hello I am Nittaya Schlamp Nice to meet you my family image

File: lib/SIL/IR/SILDeclRef.cpp (hunk 8 above)
Comment (egorzhdan) [reply lib/SIL/IR/SILDeclRef.cpp:987]: The parent function is marked as foreign because it has a Clang node, but the default argument generator isn't foreign, and doesn't require a thunk despite its parent function being foreign. This check needs to happen before the `requiresForeignToNativeThunk` check below causes us to assume that a thunk is needed.

Comment (Oudinunit) [reply lib/SIL/IR/SILDeclRef.cpp:987]: Hello How are you @egorzhdan @rjmccall @hyp @Oudinunit @lilyball @drodriguez @drodriguez @mattt @xwu @jtbandes @erica

File: lib/SILGen/SILGen.cpp (hunk 10 above)
Comment (egorzhdan) [reply lib/SILGen/SILGen.cpp]: Oh, I didn't realize we could just check the function's module context. Fixed.

File: lib/ClangImporter/ImporterImpl.h (hunk 6 above)
Comment (egorzhdan) [reply lib/ClangImporter/ImporterImpl.h]: I reverted this to `unique_ptr` and replaced with something simpler.

We're now reusing `ClangImporter::Implementation::getMangledName` to create an Itanium-mangled name for the synthesized Swift function that generates the default argument. We can't reuse `ClangImporter::Implementation::Mangler` though, because it might not be an Itanium mangler, which would produce a name that isn't a valid identifier in Swift.

File: lib/IRGen/GenClangDecl.cpp (hunk 7 above)
Comment (egorzhdan) [reply lib/IRGen/GenClangDecl.cpp:80]: We don't actually emit the generator unless it is used. Clang's CodeGen adds this decl to the list of deferred decls (https://github.com/apple/llvm-project/blob/d4738f0f52791c6004131af7322f13a07abee8e7/clang/lib/CodeGen/CodeGenModule.cpp#L3679C2-L3679C2) and emits it only if it's referenced.

I added an IRGen test to make sure unused generators aren't emitted.

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

File: lib/IRGen/GenClangDecl.cpp (hunk 7 above)
Comment (rjmccall) [reply lib/IRGen/GenClangDecl.cpp:80]: Ah, that makes sense.

**Review by rjmccall — APPROVED** [review]
LGTM

**Comment by Oudinunit** [comment]
I see you guys controlling me. Because I saw my profile overlapping 1 profile on GitHub. in my phone. I normally only have 2 profiles. But I have 3 profiles now. Because I don’t have profile me. again. you  know my family @egorzhdan
